File: 백준/Silver/4963. 섬의 개수/섬의 개수.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.LinkedList;
import java.util.Queue;

public class Main {

    public static void bfs(int x, int y, int[][] island, boolean[][] visited, int w, int h) {
        int[] dx = {-1, 1, 0, 0, -1, -1, 1, 1};
        int[] dy = {0, 0, -1, 1, -1, 1, -1, 1};

        Queue<int[]> q = new LinkedList<>();
        q.add(new int[]{x, y});
        visited[y][x] = true;

        while (!q.isEmpty()) {
            int[] current = q.poll();
            int cx = current[0];
            int cy = current[1];

            for (int i = 0; i < 8; i++) {
                int nx = cx + dx[i];
                int ny = cy + dy[i];

                if (nx >= 0 && nx < w && ny >= 0 && ny < h && !visited[ny][nx] && island[ny][nx] == 1) {
                    visited[ny][nx] = true;
                    q.add(new int[]{nx, ny});
                }
            }
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        while (true) {
            String[] dimensions = br.readLine().split(" ");
            int w = Integer.parseInt(dimensions[0]);
            int h = Integer.parseInt(dimensions[1]);

            if (w == 0 && h == 0) {
                break;
            }

            int[][] island = new int[h][w];
            boolean[][] visited = new boolean[h][w];

            for (int i = 0; i < h; i++) {
                String[] row = br.readLine().split(" ");
                for (int j = 0; j < w; j++) {
                    island[i][j] = Integer.parseInt(row[j]);
                }
            }

            int count = 0;

            for (int y = 0; y < h; y++) {
                for (int x = 0; x < w; x++) {
                    if (island[y][x] == 1 && !visited[y][x]) {
                        bfs(x, y, island, visited, w, h);
                        count++;
                    }
                }
            }

            System.out.println(count);
        }
    }
}

File: 백준/Silver/4963. 섬의 개수/섬의 개수.py
import sys
from collections import deque
input = sys.stdin.readline
def bfs(x,y,island,visited, w,h):
    directions = [(-1, 0), (1, 0), (0, -1), (0, 1), (-1, -1), (-1, 1), (1, -1), (1, 1)]
    q = deque([(x,y)])
    visited[y][x] = True
    while q:
        cx,cy = q.popleft()
        for dx,dy in directions:
            nx, ny = cx+dx, cy+dy
            if 0<=nx <w and 0<=ny<h and not visited[ny][nx] and island[ny][nx] ==1:
                visited[ny][nx] = True
                q.append((nx,ny))

while True:
    w, h = map(int,input().split())
    island = []
    if w==0 and h==0:
        break
    for _ in range(h):
        island.append(list(map(int,input().split())))
    cnt = 0
    visited = [[False for _ in range(w)]for _ in range(h)]

    #섬 탐색
    for y in range(h):
        for x in range(w):
            if island[y][x] ==1 and not visited[y][x]:
                bfs(x,y,island,visited, w,h)
                cnt +=1

    print(cnt)